### PR TITLE
strings: add string/slice node, with tests.

### PR DESCRIPTION
**Pull request by glima**
Signed-off-by: Gustavo Lima Chaves <gustavo.lima.chaves@intel.com>

### CODE REVIEW

File: src/modules/flow/string/string-ascii.c [hunk 1]
```diff
@@ -94,6 +94,8 @@ get_string_by_port(const struct sol_flow_packet *packet,
     free(mdata->string[port]);
 
     mdata->string[port] = strdup(in_value);
+    if (!mdata->string[port])
```
Comment (barbieri) [inline src/modules/flow/string/string-ascii.c]: `SOL_NUL_CHECK()` to make it verbose on errors?

File: src/modules/flow/string/string-ascii.c [hunk 2]
```diff
@@ -236,6 +238,132 @@ string_compare(struct sol_flow_node *node,
...
+    struct sol_str_slice slice;
+    int32_t start = mdata->idx[0], end = mdata->idx[1],
+        len = strlen(mdata->str);
+
+#define BOUND_CHECK(_val) \
+    do { \
+        if (_val >= 0) { \
+            if (!len && _val) \
+                _val = 0; \
+            else if (_val > len) \
```
Comment (barbieri) [inline src/modules/flow/string/string-ascii.c]: This is the generic version of the previous test. Keep only this one.

File: src/modules/flow/string/string-ascii.c (hunk 1 above)
Comment (glima) [inline src/modules/flow/string/string-ascii.c]: Next commit.

File: src/modules/flow/string/string-ascii.c (hunk 2 above)
Comment (glima) [inline src/modules/flow/string/string-ascii.c]: Fair, thanks.

File: src/modules/flow/string/string-ascii.c [hunk 3]
```diff
@@ -236,6 +238,132 @@ string_compare(struct sol_flow_node *node,
...
+                _val = len; \
+        } else if (_val < -len) { \
+            _val = 0; \
+        } else { \
+            _val = len + _val; \
+        } \
+    } while (0)
```
Comment (barbieri) [inline src/modules/flow/string/string-ascii.c]: I'd say don't create a macro, but a `static inline` function in `sol-util.h` to clamp a number within 2 values.

Then you should compute the value and clamp it:
```c
static inline int32_t
sol_util_int32_clamp(int32_t start, int32_t end, int32_t value)
{
    if (value < start)
        return start;
    if (value > end)
        return end;
    return value;
}

static int
slice_do(...) {
    ...
    if (start < 0) start = len + start;
    if (end < 0) end = len + end;
    start = sol_util_int32_clamp(0, len, start);
    end = sol_util_int32_clamp(0, len, end);
```

Comment (glima) [inline src/modules/flow/string/string-ascii.c]: Okay, I'll test with that.

**Comment by glima** [comment]
v2 coming.

**Pull request closed without merging**